yarn/src/main/scala/org/apache/spark/deploy/yarn/AMDelegationTokenRenewer.scala

```diff
@@ -0,0 +1,211 @@
+          System.currentTimeMillis())).toLong
+        // If only 6 hours left, then force a renewal immediately. This is to avoid tokens with
+        // very less validity being used on AM restart.
+        if ((interval millis).toHours <= 6) {
+          0L
+        } else {
+          interval
+        }
+      }
+
+      def scheduleRenewal(runnable: Runnable): Unit = {
+        val renewalInterval = getRenewalInterval
+        logInfo(s"Scheduling login from keytab in $renewalInterval millis.")
+        delegationTokenRenewer.schedule(runnable, renewalInterval, TimeUnit.MILLISECONDS)
+      }
+
+      // This thread periodically runs on the driver to update the delegation tokens on HDFS.
+      val driverTokenRenewerRunnable =
+        new Runnable {
+          override def run(): Unit = {
+            var wroteNewFiles = false
+            try {
+              writeNewTokensToHDFS(principal, keytab)
+              wroteNewFiles = true
+              cleanupOldFiles()
+            } catch {
+              case e: Exception =>
+                // If the exception was due to some issue deleting files, don't worry about it -
+                // just try to clean up next time. Else, reschedule for an hour later so new
+                // tokens get written out.
+                if (!wroteNewFiles) {
+                  logWarning("Failed to write out new credentials to HDFS, will try again in an " +
+                    "hour! If this happens too often tasks will fail.", e)
+                  delegationTokenRenewer.schedule(this, 1, TimeUnit.HOURS)
+                  return
+                } else {
+                  logWarning("Error while attempting to clean up old delegation token files. " +
+                    "Cleanup will be reattempted the next time new tokens are being written.")
+                }
+            }
+            scheduleRenewal(this)
+          }
+        }
+      // Schedule update of credentials. This handles the case of updating the tokens right now
+      // as well, since the renenwal interval will be 0, and the thread will get scheduled
+      // immediately.
+      scheduleRenewal(driverTokenRenewerRunnable)
+
+    }
+  }
+
+  // Keeps only files that are newer than 30 days, and deletes everything else. At least 5 files
+  // are kept for safety
+  private def cleanupOldFiles(): Unit = {
+    import scala.concurrent.duration._
+    try {
+      val remoteFs = FileSystem.get(hadoopConf)
+      val credentialsPath = new Path(sparkConf.get("spark.yarn.credentials.file"))
+      hadoopUtil.listFilesSorted(
+        remoteFs, credentialsPath.getParent, credentialsPath.getName, ".tmp").dropRight(5)
+        .takeWhile(_.getModificationTime < System.currentTimeMillis() - (30 days).toMillis)
+        .foreach(x => remoteFs.delete(x.getPath, true))
+    } catch {
+      // Such errors are not fatal, so don't throw. Make sure they are logged though
+      case e: Exception =>
```

Comment:
So this means the `wroteNewFiles` check above is not needed, since this method won't ever throw.